File: space/paint.py
"""Fail-closed organ paint bind.

The HUD is a function of the gate. Missing cycle JSON is UNAVAILABLE, not green.
HARD_DENY / LAMBDA_VETO / DENY_DEFAULT / ESCALATE cannot paint ALLOW.
Lambda stays Conjecture 1. productionPromotion cannot authorize chrome.
Arithmetic-would-allow is a ghost and cannot enable a control.
"""
from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Callable

CYCLE_SCHEMA = "szl.frontier.ouroboros-cycle.v1"
ORGAN_CYCLE_URL = (
    "https://szlholdings-szl-frontier.hf.space/frontier/ouroboros-cycle.v1.json"
)
CYCLE_PATH = "/frontier/ouroboros-cycle.v1.json"
ORGAN_CYCLE_TIMEOUT_SECONDS = 5
DENY_VERDICTS = frozenset(
    {"HARD_DENY", "LAMBDA_VETO", "DENY_DEFAULT", "ESCALATE"}
)

Paint = str
UrlOpen = Callable[..., Any]


def paint_from_cycle(cycle: dict[str, Any] | None) -> Paint:
    if not isinstance(cycle, dict):
        return "UNAVAILABLE"
    if cycle.get("schema") != CYCLE_SCHEMA:
        return "UNAVAILABLE"
    if cycle.get("productionPromotion") is True:
        return "DENY"
    if cycle.get("lambda") != "CONJECTURE_1" or cycle.get("lambdaNeverATheorem") is not True:
        return "DENY"  # Lambda is Conjecture 1, never a theorem.
    if cycle.get("authority") != "PROPOSAL_ONLY":
        return "DENY"
    if cycle.get("invariantsOk") is not True:
        return "DENY"
    shadow = cycle.get("shadow")
    if isinstance(shadow, dict) and shadow.get("executable") is True:
        return "DENY"
    if cycle.get("verdict") == "ALLOW":
        return "ALLOW"
    if cycle.get("verdict") in DENY_VERDICTS:
        return "DENY"
    return "UNAVAILABLE"


def allow_chrome(paint: Paint) -> bool:
    return paint == "ALLOW"


def paint_tone(paint: Paint) -> str:
    if paint == "ALLOW":
        return "allow"
    if paint == "DENY":
        return "deny"
    return "pending"


def paint_label(paint: Paint, verdict: str | None = None) -> str:
    if paint == "UNAVAILABLE":
        return "UNAVAILABLE"
    if paint == "DENY":
        return str(verdict or "DENY")
    return "ALLOW"


def load_published_cycle(
    *,
    url: str = ORGAN_CYCLE_URL,
    timeout: float = ORGAN_CYCLE_TIMEOUT_SECONDS,
    urlopen: UrlOpen | None = None,
) -> dict[str, Any] | None:
    opener = urlopen or urllib.request.urlopen
    try:
        request = urllib.request.Request(
            url,
            method="GET",
            headers={"Accept": "application/json", "Cache-Control": "no-store"},
        )
        with opener(request, timeout=timeout) as response:
            status = getattr(response, "status", None)
            if status not in (None, 200):
                return None
            raw = response.read()
        body = json.loads(raw.decode("utf-8"))
    except (
        OSError,
        ValueError,
        TypeError,
        json.JSONDecodeError,
        UnicodeError,
        urllib.error.URLError,
        TimeoutError,
    ):
        return None
    if not isinstance(body, dict):
        return None
    if body.get("schema") != CYCLE_SCHEMA:
        return None
    return body

File: space/assets/paint.js
/* Fail-closed organ paint bind.
 * The HUD is a function of the gate. Missing cycle JSON is UNAVAILABLE.
 * HARD_DENY / LAMBDA_VETO / DENY_DEFAULT / ESCALATE cannot paint ALLOW.
 * Lambda stays Conjecture 1. Arithmetic-would-allow is a ghost.
 */

export const CYCLE_SCHEMA = 'szl.frontier.ouroboros-cycle.v1';
export const ORGAN_CYCLE_URL = 'https://szlholdings-szl-frontier.hf.space/frontier/ouroboros-cycle.v1.json';
export const CYCLE_PATH = '/frontier/ouroboros-cycle.v1.json';

export function paintFromCycle(cycle) {
  if (!cycle) return 'UNAVAILABLE';
  if (cycle.schema !== CYCLE_SCHEMA) return 'UNAVAILABLE';
  if (cycle.productionPromotion === true) return 'DENY';
  if (cycle.lambda !== 'CONJECTURE_1' || cycle.lambdaNeverATheorem !== true) return 'DENY';
  if (cycle.authority !== 'PROPOSAL_ONLY') return 'DENY';
  if (cycle.invariantsOk !== true) return 'DENY';
  if (cycle.shadow && cycle.shadow.executable === true) return 'DENY';
  if (cycle.verdict === 'ALLOW') return 'ALLOW';
  if (
    cycle.verdict === 'HARD_DENY' ||
    cycle.verdict === 'LAMBDA_VETO' ||
    cycle.verdict === 'DENY_DEFAULT' ||
    cycle.verdict === 'ESCALATE'
  ) {
    return 'DENY';
  }
  return 'UNAVAILABLE';
}

export function allowChrome(paint) {
  return paint === 'ALLOW';
}

export function paintTone(paint) {
  if (paint === 'ALLOW') return 'allow';
  if (paint === 'DENY') return 'deny';
  return 'pending';
}

export function paintLabel(paint, verdict) {
  if (paint === 'UNAVAILABLE') return 'UNAVAILABLE';
  if (paint === 'DENY') return String(verdict || 'DENY');
  return 'ALLOW';
}

export async function loadPublishedCycle(fetchImpl = fetch) {
  try {
    const response = await fetchImpl(CYCLE_PATH, { method: 'GET', cache: 'no-store' });
    if (!response.ok) return null;
    const body = await response.json();
    if (!body || typeof body !== 'object') return null;
    if (body.schema !== CYCLE_SCHEMA) return null;
    return body;
  } catch {
    return null;
  }
}
